core/src/main/scala/io/projectglow/sql/expressions/NormalizeVariantExpr.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright 2019 The Glow Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.projectglow.sql.expressions
+
+import java.nio.file.Paths
+
+import htsjdk.samtools.reference.IndexedFastaSequenceFile
+import org.apache.spark.TaskContext
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
+import org.apache.spark.sql.catalyst.expressions.{Expression, ImplicitCastInputTypes, SenaryExpression}
+import org.apache.spark.sql.types._
+
+object NormalizeVariantExpr {
+
+  val state = new ThreadLocal[IndexedFastaSequenceFile]
+
+  def doVariantNormalization(contigName: Any,
+                             start: Any,
+                             end: Any,
+                             refAllele: Any,
+                             altAlleles: Any,
+                             refGenomePathString: Any): InternalRow = {
+
+    if (state.get() == null) {
+      // Save fasta sequence file
+      val refGenomeIndexedFasta = new IndexedFastaSequenceFile(Paths.get(refGenomePathString.asInstanceOf[String]))
+      state.set(refGenomeIndexedFasta)
+      TaskContext.get().addTaskCompletionListener(_ => state.remove())
+    }
+
+    VariantNormalizer.normalizeVariant(
+      contigName.asInstanceOf[String],
+      start.asInstanceOf[Int],
+      end.asInstanceOf[Int],
+      refAllele.asInstanceOf[String],
+      altAlleles.asInstanceOf[Array[String]],
+      state.get()
+    )
+  }
+}
+
+
+case class NormalizeVariant(contigName: Expression,
+                            start: Expression,
+                            end: Expression,
+                            refAllele : Expression,
+                            altAlleles: Expression,
+                            refGenomePathString: Expression)
+  extends SenaryExpression
+    with ImplicitCastInputTypes {
+
+  override def prettyName: String = "normalize_variant"
+
+  override def dataType: DataType = VariantNormalizer.normalizationSchema
+
+  override def inputTypes: Seq[DataType] =
```

Comment:
I think `start` and `end` are usually longs?

Reply:
WIP effect. fixed.